// @shopify/shopify-api/dist/esm/lib/auth/index.mjs
import { callback, begin } from './oauth/oauth.mjs';
import { nonce } from './oauth/nonce.mjs';
import { safeCompare } from './oauth/safe-compare.mjs';
import { buildEmbeddedAppUrl, getEmbeddedAppUrl } from './get-embedded-app-url.mjs';
import { tokenExchange } from './oauth/token-exchange.mjs';
import { clientCredentials } from './oauth/client-credentials.mjs';
import { refreshToken } from './oauth/refresh-token.mjs';
import { migrateToExpiringToken } from './oauth/migrate-to-expiring-token.mjs';

function shopifyAuth(config) {
    const shopify = {
        begin: begin(config),
        callback: callback(config),
        nonce,
        safeCompare,
        getEmbeddedAppUrl: getEmbeddedAppUrl(config),
        buildEmbeddedAppUrl: buildEmbeddedAppUrl(config),
        tokenExchange: tokenExchange(config),
        migrateToExpiringToken: migrateToExpiringToken(config),
        refreshToken: refreshToken(config),
        clientCredentials: clientCredentials(config),
    };
    return shopify;
}

export { shopifyAuth };
