// @shopify/shopify-api/dist/cjs/lib/auth/oauth/migrate-to-expiring-token.js
'use strict';

var common = require('../../clients/common.js');
var types = require('../../clients/types.js');
var fetchRequest = require('../../utils/fetch-request.js');
var shopValidator = require('../../utils/shop-validator.js');
var createSession = require('./create-session.js');
var tokenExchange = require('./token-exchange.js');

const TokenExchangeGrantType = 'urn:ietf:params:oauth:grant-type:token-exchange';
function migrateToExpiringToken(config) {
    return async ({ shop, nonExpiringOfflineAccessToken, }) => {
        const body = {
            client_id: config.apiKey,
            client_secret: config.apiSecretKey,
            grant_type: TokenExchangeGrantType,
            subject_token: nonExpiringOfflineAccessToken,
            subject_token_type: tokenExchange.RequestedTokenType.OfflineAccessToken,
            requested_token_type: tokenExchange.RequestedTokenType.OfflineAccessToken,
            expiring: '1',
        };
        const cleanShop = shopValidator.sanitizeShop(config)(shop, true);
        const postResponse = await fetchRequest.fetchRequestFactory(config)(`https://${cleanShop}/admin/oauth/access_token`, {
            method: 'POST',
            body: JSON.stringify(body),
            headers: {
                'Content-Type': types.DataType.JSON,
                Accept: types.DataType.JSON,
            },
        });
        if (!postResponse.ok) {
            common.throwFailedRequest(await postResponse.json(), false, postResponse);
        }
        return {
            session: createSession.createSession({
                accessTokenResponse: await postResponse.json(),
                shop: cleanShop,
                // We need to keep this as an empty string as our template DB schemas have this required
                state: '',
                config,
            }),
        };
    };
}

exports.migrateToExpiringToken = migrateToExpiringToken;
