// @shopify/shopify-api/dist/cjs/lib/webhooks/types.js
'use strict';

var types$1 = require('../utils/types.js');
var types = require('../types.js');

const WebhookType = {
    Webhooks: 'webhooks',
    Events: 'events',
};
const WEBHOOK_HEADER_NAMES = {
    [WebhookType.Webhooks]: {
        hmac: types.ShopifyHeader.Hmac,
        topic: types.ShopifyHeader.Topic,
        domain: types.ShopifyHeader.Domain,
        apiVersion: types.ShopifyHeader.ApiVersion,
        webhookId: types.ShopifyHeader.WebhookId,
        subTopic: types.ShopifyHeader.SubTopic,
        name: types.ShopifyHeader.Name,
        triggeredAt: types.ShopifyHeader.TriggeredAt,
        eventId: types.ShopifyHeader.EventId,
    },
    [WebhookType.Events]: {
        hmac: types.ShopifyEventsHeader.Hmac,
        topic: types.ShopifyEventsHeader.Topic,
        domain: types.ShopifyEventsHeader.Domain,
        apiVersion: types.ShopifyEventsHeader.ApiVersion,
        webhookId: types.ShopifyEventsHeader.WebhookId,
        eventId: types.ShopifyEventsHeader.EventId,
        handle: types.ShopifyEventsHeader.Handle,
        action: types.ShopifyEventsHeader.Action,
        resourceId: types.ShopifyEventsHeader.ResourceId,
        triggeredAt: types.ShopifyEventsHeader.TriggeredAt,
    },
};
exports.DeliveryMethod = void 0;
(function (DeliveryMethod) {
    DeliveryMethod["Http"] = "http";
    DeliveryMethod["EventBridge"] = "eventbridge";
    DeliveryMethod["PubSub"] = "pubsub";
})(exports.DeliveryMethod || (exports.DeliveryMethod = {}));
// eslint-disable-next-line no-warning-comments
// TODO Rethink the wording for this enum - the operations we're doing are actually "subscribing" and "unsubscribing"
// Consider changing the values when releasing v12.0.0 when it can be safely deprecated
exports.WebhookOperation = void 0;
(function (WebhookOperation) {
    WebhookOperation["Create"] = "create";
    WebhookOperation["Update"] = "update";
    WebhookOperation["Delete"] = "delete";
})(exports.WebhookOperation || (exports.WebhookOperation = {}));
const WebhookValidationErrorReason = {
    ...types$1.ValidationErrorReason,
    MissingHeaders: 'missing_headers',
};

exports.WEBHOOK_HEADER_NAMES = WEBHOOK_HEADER_NAMES;
exports.WebhookType = WebhookType;
exports.WebhookValidationErrorReason = WebhookValidationErrorReason;
