// @shopify/api-codegen-preset/dist/preset.js
'use strict';

var graphqlCodegen = require('@shopify/graphql-codegen');
var apiConfigs = require('./helpers/api-configs.js');
var getOutputFiles = require('./helpers/get-output-files.js');

const preset = {
    buildGeneratesSection: (options) => {
        const apiType = options.presetConfig.apiType;
        const { interfaceExtension, module, presetConfigs } = apiConfigs.apiConfigs[apiType];
        // Determine if the output file is a declaration file
        const isDts = options.baseOutputDir.endsWith('.d.ts');
        // Get the correct filename with extension (.d.ts or .ts)
        const { typesFile } = getOutputFiles.getOutputFiles(apiType, isDts);
        return graphqlCodegen.preset.buildGeneratesSection({
            ...options,
            presetConfig: {
                ...presetConfigs,
                importTypes: {
                    namespace: presetConfigs.importTypes.namespace,
                    from: `./${typesFile}`,
                },
                interfaceExtension: ({ queryType, mutationType, }) => interfaceExtension
                    .replace('%%MODULE%%', options.presetConfig.module ?? module)
                    .replace('%%QUERY%%', queryType)
                    .replace('%%MUTATION%%', mutationType),
            },
        });
    },
};

exports.preset = preset;
