// @shopify/shopify-api/dist/esm/lib/webhooks/types.mjs
import { ValidationErrorReason } from '../utils/types.mjs';
import { ShopifyHeader, ShopifyEventsHeader } from '../types.mjs';

const WebhookType = {
    Webhooks: 'webhooks',
    Events: 'events',
};
const WEBHOOK_HEADER_NAMES = {
    [WebhookType.Webhooks]: {
        hmac: ShopifyHeader.Hmac,
        topic: ShopifyHeader.Topic,
        domain: ShopifyHeader.Domain,
        apiVersion: ShopifyHeader.ApiVersion,
        webhookId: ShopifyHeader.WebhookId,
        subTopic: ShopifyHeader.SubTopic,
        name: ShopifyHeader.Name,
        triggeredAt: ShopifyHeader.TriggeredAt,
        eventId: ShopifyHeader.EventId,
    },
    [WebhookType.Events]: {
        hmac: ShopifyEventsHeader.Hmac,
        topic: ShopifyEventsHeader.Topic,
        domain: ShopifyEventsHeader.Domain,
        apiVersion: ShopifyEventsHeader.ApiVersion,
        webhookId: ShopifyEventsHeader.WebhookId,
        eventId: ShopifyEventsHeader.EventId,
        handle: ShopifyEventsHeader.Handle,
        action: ShopifyEventsHeader.Action,
        resourceId: ShopifyEventsHeader.ResourceId,
        triggeredAt: ShopifyEventsHeader.TriggeredAt,
    },
};
var DeliveryMethod;
(function (DeliveryMethod) {
    DeliveryMethod["Http"] = "http";
    DeliveryMethod["EventBridge"] = "eventbridge";
    DeliveryMethod["PubSub"] = "pubsub";
})(DeliveryMethod || (DeliveryMethod = {}));
// eslint-disable-next-line no-warning-comments
// TODO Rethink the wording for this enum - the operations we're doing are actually "subscribing" and "unsubscribing"
// Consider changing the values when releasing v12.0.0 when it can be safely deprecated
var WebhookOperation;
(function (WebhookOperation) {
    WebhookOperation["Create"] = "create";
    WebhookOperation["Update"] = "update";
    WebhookOperation["Delete"] = "delete";
})(WebhookOperation || (WebhookOperation = {}));
const WebhookValidationErrorReason = {
    ...ValidationErrorReason,
    MissingHeaders: 'missing_headers',
};

export { DeliveryMethod, WEBHOOK_HEADER_NAMES, WebhookOperation, WebhookType, WebhookValidationErrorReason };
